sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -275,8 +276,37 @@ object LogicalPlanIntegrity {
    * Some plan transformers (e.g., `RemoveNoopOperators`) rewrite logical
    * plans based on this assumption.
    */
-  def checkIfExprIdsAreGloballyUnique(plan: LogicalPlan): Boolean = {
-    checkIfSameExprIdNotReused(plan) && hasUniqueExprIdsForOutput(plan)
+  def validateExprIdUniqueness(plan: LogicalPlan): Unit = {
+    if (!LogicalPlanIntegrity.checkIfSameExprIdNotReused(plan)) {
+      throw SparkException.internalError("Cannot reuse the exprId in Alias")
+    }
+    if (!LogicalPlanIntegrity.hasUniqueExprIdsForOutput(plan)) {
+      throw SparkException.internalError(
+        "Some Attributes have the same exprId but different data types")
+    }
+  }
+
+  /**
+   * Validate the structural integrity of an optimized plan.
+   * Currently we check after the execution of each rule if a plan:
+   * - is still resolved
+   * - only host special expressions in supported operators
+   * - has globally-unique attribute IDs
+   * - optimized plan have same schema with previous plan.
+   */
+  def validateOptimizedPlan(
+      previousPlan: LogicalPlan,
+      currentPlan: LogicalPlan): Unit = {
+    if (!currentPlan.resolved) {
+      throw SparkException.internalError("The plan becomes unresolved")
```

Comment:
optional: should we put a debug string of the plan in here? (not sure if we should or not, it could be quite large)